clusterloader2/api/validation/validation.go

```diff
@@ -0,0 +1,148 @@
+	allErrs := field.ErrorList{}
+	for i, phase := range s.Phases {
+		allErrs = append(allErrs, verifyPhase(&phase, fldPath.Child("phases").Index(i))...)
+	}
+	for i, measurement := range s.Measurements {
+		allErrs = append(allErrs, verifyMeasurement(&measurement, fldPath.Child("measurements").Index(i))...)
+	}
+	if (len(s.Phases) == 0) == (len(s.Measurements) == 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath, "phases, measurements", "exactly one should non-empty"))
+	}
+	return allErrs
+}
+
+// verifyPhase requires ReplicasPerNamespace to have positive value.
+func verifyPhase(p *api.Phase, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if p.NamespaceRange != nil {
+		allErrs = append(allErrs, verifyNamespaceRange(p.NamespaceRange, fldPath.Child("namespaceRange"))...)
+	}
+	if p.ReplicasPerNamespace < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "replicasPerNamespace", "must be non-negative"))
+	}
+	for i, object := range p.ObjectBundle {
+		allErrs = append(allErrs, verifyObject(&object, fldPath.Child("objectBundle").Index(i))...)
+	}
+	return allErrs
+}
+
+func verifyObject(o *api.Object, fldPath *field.Path) field.ErrorList {
+	return verifyObjectType(&o.ObjectType, fldPath.Child("objectType"))
+}
+
+// verifyNamespaceRange requires Min and Max to have positive values
+// and Max to be greater than Min.
+func verifyNamespaceRange(nr *api.NamespaceRange, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if nr.Min < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "min", "must be non-negative"))
+	}
+	if nr.Max < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "max", "must be non-negative"))
+	}
+	if nr.Max < nr.Min {
+		allErrs = append(allErrs, field.Invalid(fldPath, "min, max", "Max must be greater than Min"))
+	}
+	return allErrs
+}
+
+func verifyObjectType(ot *api.ObjectType, fldPath *field.Path) field.ErrorList {
+	return field.ErrorList{}
+}
+
+// verifyTuningSet requires exactly one of tuning set type structure
+// (QpsLoad, RandomizedLoad or SteppedLoad) to be non-emtpy.
+func verifyTuningSet(ts *api.TuningSet, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	setTypes := 0
+	if ts.QpsLoad != nil {
+		allErrs = append(allErrs, verifyQpsLoad(ts.QpsLoad, fldPath.Child("qpsLoad"))...)
+		setTypes += 1
+	}
+	if ts.RandomizedLoad != nil {
+		allErrs = append(allErrs, verifyRandomizedLoad(ts.RandomizedLoad, fldPath.Child("randomizedLoad"))...)
+		setTypes += 1
+	}
+	if ts.SteppedLoad != nil {
+		allErrs = append(allErrs, verifySteppedLoad(ts.SteppedLoad, fldPath.Child("steppedLoad"))...)
+		setTypes += 1
+	}
+	if setTypes != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "qpsLoad, randomizedLoad, steppedLoad", "exactly one type of TuningSet should be set"))
+	}
+	return allErrs
+}
+
+func verifyMeasurement(m *api.Measurement, fldPath *field.Path) field.ErrorList {
+	return field.ErrorList{}
```

Comment:
return nil (same everywhere below in such case)

Reply:
done